flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/MailboxExecutorImpl.java

```diff
@@ -37,9 +39,12 @@
 
 	private final int priority;
 
-	public MailboxExecutorImpl(@Nonnull TaskMailbox mailbox, int priority) {
+	public final ExecutionDecorator executionDecorator;
+
+	public MailboxExecutorImpl(@Nonnull TaskMailbox mailbox, int priority, ExecutionDecorator executionDecorator) {
```

Comment:
maybe instead of wrapping every `#run()` with the decorator, we could provide
```
public class DecoratedMail extends Mail {
  private final ExecutionDecorator executionDecorator;
  (...)
  @Override
  public void run() {
    executionDecorator.runThrowing(super::run);
  }
}
```
? After all `Mail` is constructed only in two places in the `MailboxExecutorImpl`, which could be deduplicated into a single `Mail MailboxExecutorImpl#createMail(...)`. From here we could either ad `if` check inside this method that would created `DecoratedMail` or `Mail`, or introduce `DecoratedMailboxExecutor` which would override this method? (independent `DecoratedMailboxExecutor` would ensure no extra overheads when not using checkpoint lock)

As it is now, it's a bit fragile to remember that any mail execution should happen with using `executionDecorator`.

Reply:
Not all wrapped actions aren't `Mails`, so we will end up either wrapping them too into `Mail` or having different approaches to execute `Mails` and regular actions.